### DISCUSSION TITLE
Access token size with admin rights on multiple realms

### DISCUSSION THREAD

**Discussion by devneutz** [keycloak/keycloak#21549] [Q&A] [answered]
In order to let powerusers set up new tenants for our multi-tenant cloud application, we built a custom frontend (let's call it `admin-ui`). This is backed by a separate backend service, holding all the tenant and product information (let's call it `admin-api`). 

Whenever a poweruser creates a new tenant, the `admin-api` takes care of setting up the realm within keycloak (clients, idps, users, roles etc.).

The architecture can thus be described as a service chaining in its simplest form: `admin-ui` > `admin-api` > `keycloak-api`. Each service is authenticated by keycloak as the auth-server.

Whilst we have a client for the `admin-ui` using PKCE, we are reissueing tokens for `admin-api` with the client credentials grant. Behind that client credentials grant we have a service account, which holds the necessary permissions/roles for creating the recourses mentioned before.

This works perfectly fine, until we have more than a certain amount of realms. At that point, the access token gets too large for the browser and/or NGINX to handle. Unfortunately omitting the roles from the default scope is not working, the `keycloak-api` is returning a 403 Forbidden.

There are multiple different ways of tackling this issue:
- Dead man delegation of an admin token issued by admin-ui
- Increase of buffer-size to support larger access tokens (not scalable)
- Decrease number of roles for service account (not scalable)

I am wondering what the best practice would be? Is this even the right way of going at it or is there a better solution (different grant type, different architecture)?

**Comment by danielFesenmeyer** [comment, marked as answer]
Can't you use the "admin-cli" or "security-admin-console" client? For those clients, Keycloak reads the roles directly from Keycloak, **not** from the token. Hence, you don't need any roles in the token, it will stay small regardless of the amount of realms and associated roles.

**Reply by devneutz** [reply]
The `security-admin-console` client is not a bearer-only client, which makes it unusable in our scenario since the `admin-api` is a backend service without user interaction.

The `admin-cli` client however looks like it could work - in my understanding it would just need a user/pwd instead of a clientid/clientsecret combination, correct?

**Reply by devneutz** [reply]
On that note, what makes these clients different from custom clients? Why does keycloak read the roles directly instead from the token only if you use these 2 clients? Are there configuration options that we can use for custom clients as well?

**Reply by danielFesenmeyer** [reply]
Afaik, what makes those clients special is just that they are used by the Keycloak builtin administration functionality, where often access to more than one realm is needed (master admin access). Probably that's why those clients are handled in a special way, to avoid the issue you mentioned above.

Note there is a discussion on this topic and it seems that some progress has already been made: https://github.com/keycloak/keycloak/discussions/19649

**Reply by buntyrathod** [reply]
@devneutz did you able to get this working using `admin-cli` client? Even with `admin-cli` client it does not work unless you add roles in access_token

**Reply by devneutz** [reply]
@buntyrathod yes indeed I made it work by using the admin-cli client
